### PR TITLE
Issue 45

### LINKED ISSUE #45: Disable passphrase and enable setting not root owner on certificate directory
Some applications, like redis  =< 6.0, do not support passphrase for the private key. Also, they can not read the certificate when the root user is the owner of them.

### PR DESCRIPTION
**Pull request by afeefghannam89**
Fix #45 

These are some changes which suit the redis usage of this role.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -37,8 +37,12 @@ You need to have the Python library `cryptography` in version `>1.2.3` available
...
 * `ca_server_cert`: Create server certificate as well (default: `true`)
 * `ca_logstash`: Create Logstash compatible certificate as well. Needs `ca_server_cert` to be set. (default: `false`)
-* `ca_keypassphrase`: Password for the client key (default: `ChangeMeAgain`)
-* `ca_openssl_cipher`: Cipher to use for key creation (default: `auto`)
+* `ca_keypassphrase`: Password for the client key, default not defined
+* `ca_openssl_cipher`: Cipher to use for key creation, default not defined
+* `client_ca_dir`: Directory to place CA and certificates on the clients (default: `/opt/ca`)
+* `client_ca_dir_owner`: User to owne the certificate directory on the clients (default: `root`)
+* `client_ca_dir_group`: Group to owne the certificate directory on the clients (default: `root`)
+* `client_ca_dir_mode`: Permissions of the certificate directory on the clients (default: `0700`)
```
Comment (widhalmt) [inline README.md:45]:
```suggestion
* `ca_client_ca_dir`: Directory to place CA and certificates on the clients (default: `/opt/ca`)
* `ca_client_ca_dir_owner`: User to own the certificate directory on the clients (default: `root`)
* `ca_client_ca_dir_group`: Group to own the certificate directory on the clients (default: `root`)
* `ca_client_ca_dir_mode`: Permissions of the certificate directory on the clients (default: `0700`)
```

Please prefix variables with the name of the role. And a small typo. :-)

File: defaults/main.yml [hunk 2]
```diff
@@ -2,6 +2,10 @@
 
 ca_manage_openssl: true
 ca_ca_dir: /opt/ca
+client_ca_dir: /opt/ca
+client_ca_dir_owner: root
+client_ca_dir_group: root
+client_ca_dir_mode: 0700
```
Comment (widhalmt) [inline defaults/main.yml:8]:
```suggestion
ca_client_ca_dir: /opt/ca
ca_client_ca_dir_owner: root
ca_client_ca_dir_group: root
ca_client_ca_dir_mode: 0700
```
Same renaming.

File: defaults/main.yml [hunk 3]
```diff
@@ -24,5 +28,3 @@ ca_altname_2: "{{ ansible_fqdn }}"
 ca_altname_3: "{{ inventory_hostname }}"
 ca_ca_keylength: 2048
```
Comment (widhalmt) [inline defaults/main.yml:30]:
```suggestion
```
Extra blank line

File: tasks/main.yml [hunk 4]
```diff
@@ -1,12 +1,21 @@
 ---
 
+- name: Ensure CA directory exists on clients
+  file:
+    path: "{{ client_ca_dir }}"
+    owner: "{{ client_ca_dir_owner }}"
+    group: "{{ client_ca_dir_group }}"
+    mode: "{{ client_ca_dir_mode }}"
```
Comment (widhalmt) [inline tasks/main.yml:8]:
```suggestion
    path: "{{ ca_client_ca_dir }}"
    owner: "{{ ca_client_ca_dir_owner }}"
    group: "{{ ca_client_ca_dir_group }}"
    mode: "{{ ca_client_ca_dir_mode }}"
```

File: tasks/main.yml [hunk 5]
```diff
@@ -29,26 +38,28 @@
 
 - name: Create key
   openssl_privatekey:
-    path: "{{ ca_ca_dir }}/{{ inventory_hostname }}.key"
-    passphrase: "{{ ca_keypassphrase }}"
-    cipher: "{{ ca_openssl_cipher }}"
+    path: "{{ client_ca_dir }}/{{ inventory_hostname }}.key"
```
Comment (widhalmt) [inline tasks/main.yml:41]:
```suggestion
    path: "{{ ca_client_ca_dir }}/{{ inventory_hostname }}.key"
```

File: tasks/main.yml [hunk 6]
```diff
@@ -29,26 +38,28 @@
...
+    passphrase: "{{ ca_keypassphrase | default(omit) }}"
+    cipher: "{{ ca_openssl_cipher | default(omit) }}"
+    owner: "{{ client_ca_dir_owner }}"
+    group: "{{ client_ca_dir_group }}"
+    mode: "{{ client_ca_dir_mode }}"
```
Comment (widhalmt) [inline tasks/main.yml:46]:
```suggestion
    owner: "{{ ca_client_ca_dir_owner }}"
    group: "{{ ca_client_ca_dir_group }}"
    mode: "{{ ca_client_ca_dir_mode }}"
```

File: tasks/main.yml [hunk 7]
```diff
@@ -29,26 +38,28 @@
 
 - name: Create key
   openssl_privatekey:
-    path: "{{ ca_ca_dir }}/{{ inventory_hostname }}.key"
-    passphrase: "{{ ca_keypassphrase }}"
-    cipher: "{{ ca_openssl_cipher }}"
+    path: "{{ client_ca_dir }}/{{ inventory_hostname }}.key"
+    passphrase: "{{ ca_keypassphrase | default(omit) }}"
+    cipher: "{{ ca_openssl_cipher | default(omit) }}"
```
Comment (widhalmt) [inline tasks/main.yml:43]: I like how you offer to set these variables but don't force them on the user!

File: tasks/main.yml [hunk 8]
```diff
@@ -29,26 +38,28 @@
...
+    mode: "{{ client_ca_dir_mode }}"
   register: key
 
 - name: Create CSR
   openssl_csr:
-    path: "{{ ca_ca_dir }}/{{ inventory_hostname }}.csr"
-    privatekey_path: "{{ ca_ca_dir }}/{{ inventory_hostname }}.key"
-    privatekey_passphrase: "{{ ca_keypassphrase }}"
+    path: "{{ client_ca_dir }}/{{ inventory_hostname }}.csr"
+    privatekey_path: "{{ client_ca_dir }}/{{ inventory_hostname }}.key"
```
Comment (widhalmt) [inline tasks/main.yml:52]:
```suggestion
    path: "{{ ca_client_ca_dir }}/{{ inventory_hostname }}.csr"
    privatekey_path: "{{ ca_client_ca_dir }}/{{ inventory_hostname }}.key"
```

File: tasks/main.yml [hunk 9]
```diff
@@ -29,26 +38,28 @@
...
+    privatekey_passphrase: "{{ ca_keypassphrase | default(omit) }}"
     country_name: "{{ ca_country }}"
     organization_name: "{{ ca_organization }}"
     common_name: "{{ inventory_hostname }}"
     subject_alt_name: "DNS:{{ ansible_hostname }},DNS:{{ ansible_fqdn }},DNS:{{ inventory_hostname }}" # yamllint disable-line rule:line-length
-    extended_key_usage:
-      - clientAuth
+    extended_key_usage: "{{ extended_key_usage | default(omit) }}"
```
Comment (widhalmt) [inline tasks/main.yml:58]: Way better, thank you!

File: tasks/main.yml [hunk 10]
```diff
@@ -29,26 +38,28 @@
...
 
 - name: Pull CSR
   fetch:
-    src: "{{ ca_ca_dir }}/{{ inventory_hostname }}.csr"
+    src: "{{ client_ca_dir }}/{{ inventory_hostname }}.csr"
```
Comment (widhalmt) [inline tasks/main.yml:62]:
```suggestion
    src: "{{ ca_client_ca_dir }}/{{ inventory_hostname }}.csr"
```

File: tasks/main.yml [hunk 11]
```diff
@@ -81,10 +92,10 @@
 - name: Push certificate to client
   copy:
     src: "{{ ca_localdir }}/{{ inventory_hostname }}.crt"
-    dest: "{{ ca_ca_dir }}/{{ inventory_hostname }}.crt"
-    owner: root
-    group: root
-    mode: 0600
+    dest: "{{ client_ca_dir }}/{{ inventory_hostname }}.crt"
+    owner:  "{{ client_ca_dir_owner }}"
```
Comment (widhalmt) [inline tasks/main.yml:96]:
```suggestion
    dest: "{{ ca_client_ca_dir }}/{{ inventory_hostname }}.crt"
    owner:  "{{ ca_client_ca_dir_owner }}"
```

File: tasks/main.yml [hunk 12]
```diff
@@ -81,10 +92,10 @@
...
+    group: "{{ client_ca_dir_group }}"
+    mode: "{{ client_ca_dir_mode }}"
```
Comment (widhalmt) [inline tasks/main.yml:98]:
```suggestion
    group: "{{ ca_client_ca_dir_group }}"
    mode: "{{ ca_client_ca_dir_mode }}"
```

File: tasks/main.yml [hunk 13]
```diff
@@ -96,10 +107,10 @@
...
   copy:
     src: "{{ ca_localdir }}/ca.crt"
-    dest: "{{ ca_ca_dir }}/ca.crt"
-    owner: root
-    group: root
-    mode: 0600
+    dest: "{{ client_ca_dir }}/ca.crt"
+    owner: "{{ client_ca_dir_owner }}"
+    group: "{{ client_ca_dir_group }}"
+    mode: "{{ client_ca_dir_mode }}"
```
Comment (widhalmt) [inline tasks/main.yml:113]:
```suggestion
    dest: "{{ ca_client_ca_dir }}/ca.crt"
    owner: "{{ ca_client_ca_dir_owner }}"
    group: "{{ ca_client_ca_dir_group }}"
    mode: "{{ ca_client_ca_dir_mode }}"
```

File: tasks/main.yml [hunk 14]
```diff
@@ -108,19 +119,18 @@
 
     - name: Create server CSR
       openssl_csr:
-        path: "{{ ca_ca_dir }}/{{ inventory_hostname }}-server.csr"
-        privatekey_path: "{{ ca_ca_dir }}/{{ inventory_hostname }}.key"
-        privatekey_passphrase: "{{ ca_keypassphrase }}"
+        path: "{{ client_ca_dir }}/{{ inventory_hostname }}-server.csr"
+        privatekey_path: "{{ client_ca_dir }}/{{ inventory_hostname }}.key"
+        privatekey_passphrase: "{{ ca_keypassphrase | default(omit) }}"
```
Comment (widhalmt) [inline tasks/main.yml:124]:
```suggestion
        path: "{{ ca_client_ca_dir }}/{{ inventory_hostname }}-server.csr"
        privatekey_path: "{{ ca_client_ca_dir }}/{{ inventory_hostname }}.key"
        privatekey_passphrase: "{{ ca_keypassphrase | default(omit) }}"
```

File: tasks/main.yml [hunk 15]
```diff
@@ -108,19 +119,18 @@
...
         common_name: "{{ inventory_hostname }}"
         subject_alt_name: "DNS:{{ ansible_hostname }},DNS:{{ ansible_fqdn }},DNS:{{ inventory_hostname }}" # yamllint disable-line rule:line-length
-        extended_key_usage:
-          - serverAuth
+        extended_key_usage: "{{ extended_key_usage | default(omit) }}"
 
     - name: Pull server CSR
       fetch:
-        src: "{{ ca_ca_dir }}/{{ inventory_hostname }}-server.csr"
+        src: "{{ client_ca_dir }}/{{ inventory_hostname }}-server.csr"
```
Comment (widhalmt) [inline tasks/main.yml:133]:
```suggestion
        src: "{{ ca_client_ca_dir }}/{{ inventory_hostname }}-server.csr"
```

File: tasks/main.yml [hunk 16]
```diff
@@ -153,10 +163,10 @@
...
       copy:
         src: "{{ ca_localdir }}/{{ inventory_hostname }}-server.crt"
-        dest: "{{ ca_ca_dir }}/{{ inventory_hostname }}-server.crt"
-        owner: root
-        group: root
-        mode: 0600
+        dest: "{{ client_ca_dir }}/{{ inventory_hostname }}-server.crt"
+        owner: "{{ client_ca_dir_owner }}"
+        group: "{{ client_ca_dir_group }}"
+        mode: "{{ client_ca_dir_mode }}"
```
Comment (widhalmt) [inline tasks/main.yml:170]:
```suggestion
        dest: "{{ ca_client_ca_dir }}/{{ inventory_hostname }}-server.crt"
        owner: "{{ ca_client_ca_dir_owner }}"
        group: "{{ ca_client_ca_dir_group }}"
        mode: "{{ ca_client_ca_dir_mode }}"

```

**Review by widhalmt — CHANGES_REQUESTED** [review]
More or less my request can be summed up as "please prefix all variables with the name of the role". If there's a good reason you don't want that here, that might be ok for. e.g. when you need variables for interaction between roles.

*if* you want to accept my suggestion, you can make your life easier by just running `sed -i 's/client_ca/ca_client_ca/g' filename` on every file you changed and pushing that. I'll happily review it again.

**Comment by afeefghannam89** [comment]
@widhalmt Thanks for the hint and the help. Your suggested changes have been done.

**Review by widhalmt — APPROVED** [review]
Thanks for the code. This will make the role way more flexible!. Looks very good to me.

**Pull request merged**